src/java.base/share/classes/jdk/internal/misc/OperatingSystem.java

```diff
@@ -0,0 +1,148 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.misc;
+
+import jdk.internal.util.StaticProperty;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.util.Objects;
+
+/**
+ * Enumeration of operating system types and testing for the current OS.
+ * The enumeration can be used to dispatch to OS specific code or values.
+ * Checking if a specific operating system is current uses a simple
+ * static method for each operating system.
+ * <p>
+ * For example,
+ * {@snippet lang = "java":
+ * if (OperatingSystem.isWindows()) {
+ *     // Windows only code.
+ * } else if (OperatingSystem.isLinux()) {
+ *     // Linux only code
+ * }
+ *}
+ *
+ * Alternatively, compare with the {@linkplain #current() current} operating system.
+ * For example,
+ * {@snippet lang = "java":
+ * if (OperatingSystem.current() == Windows) {
+ *     // Windows only code.
+ * }
+ *}
+ * Dispatch based on the current operating system or choose a value.
+ * For example,
+ * {@snippet lang = "java":
+ * int port() {
+ *      return switch(OperatingSystem.current()) {
+ *          case Linux->32768;
+ *          case AIX->32768;
+ *          case Mac->49152;
+ *          case Windows->49152;
+ *      };
+ * }
+ * }
+ */
+public enum OperatingSystem {
+
+    /**
+     * The Linux Operating system.
+     */
+    Linux("Linux"),
+    /**
+     * The Mac OS X Operating system.
+     */
+    Mac("Mac OS X"),
+    /**
+     * The Windows Operating system.
+     */
+    Windows("Windows"),
+    /**
+     * The AIX Operating system.
+     */
+    AIX("AIX"),
+    ;
+
+    // Cache a copy of the array for lightweight indexing
+    private static final OperatingSystem[] osValues = OperatingSystem.values();
+
+    /**
+     * The operating system name.
+     */
+    private final String name;
+
+    /**
+     * {@return {@code true} if built for the Linux operating system}
+     */
+    @ForceInline
+    public static boolean isLinux() {
+        return OperatingSystemProps.TARGET_OS_IS_LINUX;
```

Comment:
```suggestion
        return OperatingSystemProps.CURRENT_OS_ORDINAL == Linux.ordinal();
```

This will also simplify the template file as `TARGET_OS_IS_XXX` constants are not needed.

Also suggest to rename `TARGET_OS_ORDINAL` to `CURRENT_OS_ORDINAL` since it represents the current OS (vs the build target).

Reply:
That would not yield a compile time constant.
The TARGET_IS_XXX values must evaluate to compile time constants as evaluated by javac.

Reply:
Don't we have a conditional compilation ability with these template files such that we can just generate true or false depending on the OS?

Reply:
Good point.  `OperatingSystemProps.java` can be a OS-specific class like: 
```
src/java.base/macosx/classes/jdk/internal/misc/OperatingSystemProps.java
              linux/classes/jdk/internal/misc/OperatingSystemProps.java
              windows/classes/jdk/internal/misc/OperatingSystemProps.java
```

OTOH, if we include `os.arch`, that would need to be a template.   A single template file to include both OS and ARCH would do it.

Reply:
I was literally thinking of the `#if[xx]` conditional compilation facility rather than an OS specific file, but okay ...

Reply:
The stream preprocessor (SPP) is more involved than the simple substitutions used to insert the target os into the template.